docs: document dynamic composition versions (#1087)
apollographql/apollo-cli: docs/source/supergraphs.md

  gateway --- serviceA & serviceB & serviceC;
```

Rover commands that interact with supergraphs begin with `rover supergraph`. These commands primarily deal with [supergraph schemas](/federation/federated-types/overview/), which adhere to the [core schema specification](https://specs.apollo.dev/core/#core-schemas).

## Composing a supergraph schema

You can use the `supergraph compose` command to compose a supergraph schema based on a **supergraph configuration file**, like so:

```bash
rover supergraph compose --config ./supergraph.yaml
```

### Configuration

> The **supergraph configuration file**, often referred to as `supergraph.yaml`, contains [subgraph](https://www.apollographql.com/docs/federation/subgraphs) definitions declared via YAML. A supergraph with two subgraphs, `films` and `people` are show in this example:

Reviewer: ```suggestion
The **supergraph configuration file** (often referred to as `supergraph.yaml`) includes configuration options for each of your [subgraphs](https://www.apollographql.com/docs/federation/subgraphs). The following example file configures a supergraph with two subgraphs, `films` and `people`:
```

  gateway --- serviceA & serviceB & serviceC;
```

> Rover commands that interact with supergraphs begin with `rover supergraph`. These commands primarily deal with [supergraph schemas](/federation/federated-types/overview/), which adhere to the [core schema specification](https://specs.apollo.dev/core/v0.2/).

Reviewer 2: **suggestion** I think it'll be more evergreen to link to technical specification, which will be able to update the version more easily

```suggestion
Rover commands that interact with supergraphs begin with `rover supergraph`. These commands primarily deal with [supergraph schemas](/federation/federated-types/overview/), which adhere to the [core schema specification](https://specs.apollo.dev/#core-schemas).
```

> Rover commands that interact with supergraphs begin with `rover supergraph`. These commands primarily deal with [supergraph schemas](/federation/federated-types/overview/). These schemas adheres to the [core schema specification](https://specs.apollo.dev/core/v0.2/) instead of the GraphQL specification.

Reviewer: ```suggestion
Rover commands that interact with supergraphs begin with `rover supergraph`. These commands primarily deal with [supergraph schemas](/federation/federated-types/overview/), which adhere to the [core schema specification](https://specs.apollo.dev/core/v0.2/) in addition to the GraphQL specification.
```

... right? Supergraph schemas don't violate the main GraphQL spec or anything, do they?

Author: uhhhh you're probably right here. i was under the assumption that it was different because of `rover graph fetch` and `rover supergraph fetch`, but you're likely right here and federation is mostly just a standard implementation of a set of directives? as usual, writing the docs highlights the fact that my understanding is fuzzier than i realized.

      subgraph: actors
```

> #### Using Federation 2

Reviewer: ```suggestion
#### Using with Federation 2
```

I suggest also adding `federation_version: 2` to the YAML example just above here

> In order to use Federation 2, you will need to put `federation_version: 2` in your `supergraph.yaml`.

Reviewer: ```suggestion
To use Rover with a Federation 2 supergraph, you need to add `federation_version: 2` to your YAML configuration file.
```

> The first time you run `rover supergraph compose` with Federation 2 on a machine, you will need to accept the terms and conditions of the [ELv2 license](https://www.apollographql.com/docs/resources/elastic-license-v2-faq/). On future invocations, Rover will remember that you already accepted the license and will not prompt you again (even if you update Rover).

Reviewer: ```suggestion
The first time you run `rover supergraph compose` with Federation 2 on a particular machine, you need to accept the terms and conditions of the [ELv2 license](https://www.apollographql.com/docs/resources/elastic-license-v2-faq/). On future invocations, Rover remembers that you already accepted the license and does not prompt you again (even if you update Rover).
```

>  CI systems will wipe away any persisted configuration on each run, and they will also prevent you from being able to accept the interactive prompt. Therefore, in CI, you should set the environment variable `APOLLO_ELV2_LICENSE=accept` or run `rover supergraph compose --config ./supergraph.yaml --elv2-license accept` to ensure that composition works smoothly.

Reviewer: ```suggestion
 > ⚠️ **Important:** CI systems wipe away any persisted Rover configuration on each run, and they can't accept the interactive ELv2 prompt. To automatically accept the prompt in CI, do one of the following:
>
> * Set the environment variable `APOLLO_ELV2_LICENSE=accept` in your CI environment.
> * Include `--elv2-license accept` in your `rover supergraph compose` command.
```

      subgraph: actors
```

#### Using with Federation 2

To use Rover with a Federation 2 supergraph, you need to add `federation_version: 2` to your YAML configuration file.

> ```yaml

Reviewer: ```suggestion
```yaml {1}
```

Reviewer: (This highlights the relevant line)

federation_version: 2
subgraphs:
  films:
    routing_url: https://films.example.com
    schema:
      file: ./films.graphql
```

> The first time you run `rover supergraph compose` with Federation 2 on a particular machine, you will need to accept the terms and conditions of the [ELv2 license](https://www.apollographql.com/docs/resources/elastic-license-v2-faq/). On future invocations, Rover remembers that you already accepted the license and will not prompt you again (even if you update Rover).

Reviewer: ```suggestion
The first time you run `rover supergraph compose` with Federation 2 on a particular machine, you're prompted to accept the terms and conditions of the [ELv2 license](https://www.apollographql.com/docs/resources/elastic-license-v2-faq/). On future invocations, Rover remembers that you already accepted the license and doesn't prompt you again (even if you update Rover).
```

> ⚠️ **Important:** CI systems wipe away any persisted Rover configuration on each run, and they can't accept the interactive ELv2 prompt. To automatically accept the prompt in CI, do one of the following:
>
> * Set the environment variable `APOLLO_ELV2_LICENSE=accept` in your CI environment.
> * Include `--elv2-license accept` in your `rover supergraph compose` command.
> * Run `yes | rover supergraph compose`

### Output format

> By default, `rover supergraph compose` outputs a [supergraph schema](/federation/federated-types/overview/) document to `stdout`. This artifact is used to start up [`@apollo/gateway`](/federation/api/apollo-gateway/) or the [🦀 Apollo Router](/router).

Reviewer: ```suggestion
By default, `rover supergraph compose` outputs a [supergraph schema](/federation/federated-types/overview/) document to `stdout`. You provide this artifact to [`@apollo/gateway`](/federation/api/apollo-gateway/) or the [🦀 Apollo Router](/router/) on startup.
```

#### Gateway compatibility

> *Apollo Gateway fails to start up if it's provided with a supergraph schema that it doesn't support. To ensure compatibility, we recommend that you test launching your gateway in a CI pipeline with the supergraph schema it will ultimately use in production.*

Reviewer: ```suggestion
**Apollo Gateway fails to start up if it's provided with a supergraph schema that it doesn't support.** To ensure compatibility, we recommend that you test launching your gateway in a CI pipeline with the supergraph schema it will ultimately use in production.
```

> Prior to Rover v0.5.0, `rover supergraph compose` shipped with exactly one version of composition that was compatible with Federation 1. This function was sourced from the [`@apollo/federation`](https://www.npmjs.com/package/@apollo/federation) JavaScript package. Therefore, it was important to keep track of your Rover version and your Gateway version and keep them in sync according to the following compatibility table.

Reviewer: I'd put the prior-to-5.0 stuff at the _end_ of this section

|Rover version|Gateway version|
|---|---|
|<= v0.2.x|<= v0.38.x|
|>= v0.3.x|>= v0.39.x|

> After Rover v0.5.0, the `rover supergraph compose` command produces a supergraph schema by dynamically selecting the proper composition function from either the [`@apollo/federation`](https://www.npmjs.com/package/@apollo/federation) package for Federation 1, or from the [`@apollo/composition`](https://www.npmjs.com/package/@apollo/composition) library for Federation 2. Rover works with _multiple_ versions of composition without needing to build and release a separate version of Rover.

Reviewer: ```suggestion
In Rover v0.5.0 and later, the `rover supergraph compose` command generates a supergraph schema by dynamically selecting the proper composition function either from the [`@apollo/federation`](https://www.npmjs.com/package/@apollo/federation) package for Federation 1, or from the [`@apollo/composition`](https://www.npmjs.com/package/@apollo/composition) library for Federation 2. Rover works with _multiple_ versions of composition without needing to build and release a separate version of Rover.
```

> It is recommended that you set `federation_version: 1` or `federation_version: 2` in your `supergraph.yaml`, and Rover will download the latest compatible composition function and use that automatically. Our aim is to release only backwards compatible changes moving forward.

Reviewer: ```suggestion
It is recommended that you set `federation_version: 1` or `federation_version: 2` in your YAML configuration file. Rover downloads the latest compatible composition function and uses it automatically. Our aim is to release only backward-compatible changes moving forward.
```

> _If_ Rover updates the composition version for you automatically and the update makes composition fail, you can pin the federation version in `supergraph.yaml` to ensure that Rover uses the same version each time. For instance: `federation_version: =2.0.0` will make Rover _always_ use `@apollo/composition@v2.0.0`. Versions of `supergraph compose` are installed to `~/.rover/bin` if you installed with the `curl | sh` installer, and to `./node_modules/.bin/` if you installed with npm.

Reviewer: ```suggestion
> **If Rover updates its composition version and causes composition to fail,** you can pin the federation version in your YAML configuration file to ensure that Rover uses the same version each time. For example, setting `federation_version: =2.0.0` ensures that Rover _always_ uses `@apollo/composition@v2.0.0`.
>
>Versions of `supergraph compose` are installed to `~/.rover/bin` if you installed with the `curl | sh` installer, and to `./node_modules/.bin/` if you installed with npm.
```

> Additionally, if you set `federation_version: 1` or `federation_version: 2`, you can run `rover supergraph compose` with the `--skip-update` flag to prevent Rover from downloading newer composition versions. This can be helpful if you are on a slow network.

Reviewer: ```suggestion
If you set `federation_version: 1` or `federation_version: 2`, you can run `rover supergraph compose` with the `--skip-update` flag to prevent Rover from downloading newer composition versions. This can be helpful if you're on a slow network.
```

#### Gateway compatibility

**Apollo Gateway fails to start up if it's provided with a supergraph schema that it doesn't support.** To ensure compatibility, we recommend that you test launching your gateway in a CI pipeline with the supergraph schema it will ultimately use in production.

> The `rover supergraph compose` command generates a supergraph schema by dynamically selecting the proper composition function either from the [`@apollo/federation`](https://www.npmjs.com/package/@apollo/federation) package for Federation 1, or from the [`@apollo/composition`](https://www.npmjs.com/package/@apollo/composition) library for Federation 2. Rover works with _multiple_ versions of composition without needing to build and release a separate version of Rover.

Reviewer 2: **suggestion** Is it more accurate to say that it selects the proper composition function based on setting `federation_version`? Dynamically selecting seems like it'll do the right thing without a setting.

```suggestion
The `rover supergraph compose` command generates a supergraph schema by selecting the proper composition function based on the `federation_version` value in your YAML configuration file. Rover will use the [`@apollo/federation`](https://www.npmjs.com/package/@apollo/federation) library for Federation 1 and the [`@apollo/composition`](https://www.npmjs.com/package/@apollo/composition) library for Federation 2. If `federation_version` is unset, Rover defaults to using Federation 1. Rover works with _multiple_ versions of composition without needing to build and release a separate version of Rover.
```

Author: sure. it is _kinda_ dynamic, since if you set it to `2` it will grab the latest in the `2` channel and if you set it to `1` it will grab the latest in the `1` channel. maybe not necessary to include the dynamic word though

Reviewer 2: There's probably an awesome buzzword for something that's quasi-dynamic

Author: completely unrelated to this, but related to awesome buzzwords: i learned that [epigraph](https://en.wikipedia.org/wiki/Epigraph_(mathematics)) is a synonym for supergraph in mathematics

Reviewer 2: 🧠

It is recommended that you set `federation_version: 1` or `federation_version: 2` in your YAML configuration file. Rover downloads the latest compatible composition function and uses it automatically. Our aim is to release only backward-compatible changes moving forward.

> > **If Rover updates its composition version and causes composition to fail,** you can pin the federation version in your YAML configuration file to ensure that Rover uses the same version each time. For example, setting `federation_version: =2.0.0` ensures that Rover _always_ uses `@apollo/composition@v2.0.0`.

Reviewer 2: **suggestion** There might be other reasons to pin a version other than composition _failure_. I wonder if we'd wanna suggest that if they _want_ to use a specific version, they can do so. Though I don't know if it's watering it down too much.

```suggestion
> **If you need to use a specific composition version,** you can pin the federation version in your YAML configuration file to ensure that Rover uses the same version each time. For example, setting `federation_version: =2.0.1` ensures that Rover _always_ uses `@apollo/composition@v2.0.1`.
```

Author: sure, i think that makes sense.

just for my reasoning behind the hedging language here - my desire is to push people to just using "latest-1" or "latest-2" and only pinning if they really need to. this is essentially the approach that managed federation takes - the studio team updates your composition function automatically for things like `subgraph publish` and `subgraph check` and there is not a way to pin the version down further. i really like the ability to deliver security updates seamlessly and pinning versions does prevent this. 

That being said I don't think there's anything wrong with just saying that they can pin the version as they please just like you've outlined here. I think it's a perfectly reasonable thing to want to pin your versions down and update them on your own, and managed federation is not yet integrated with `supergraph compose`.

Reviewer 2: That reasoning definitely makes sense. We could also warn people of the repercussions, I definitely see the danger of pinning old versions.

#### Gateway compatibility

**Apollo Gateway fails to start up if it's provided with a supergraph schema that it doesn't support.** To ensure compatibility, we recommend that you test launching your gateway in a CI pipeline with the supergraph schema it will ultimately use in production.

> The `rover supergraph compose` command generates a supergraph schema via [composition](https://www.apollographql.com/docs/federation/federated-types/composition/). For Federation 1, this algorithm was implemented in the [`@apollo/federation`](https://www.npmjs.com/package/@apollo/federation) package. For Federation 2, this algorithm was implemented in the [`@apollo/composition`](https://www.npmjs.com/package/@apollo/composition) package.

Reviewer: ```suggestion
The `rover supergraph compose` command generates a supergraph schema via [composition](https://www.apollographql.com/docs/federation/federated-types/composition/). For Federation 1, this algorithm was implemented in the [`@apollo/federation`](https://www.npmjs.com/package/@apollo/federation) package. For Federation 2, this algorithm is implemented in the [`@apollo/composition`](https://www.npmjs.com/package/@apollo/composition) package.
```

> `rover supergraph compose` supports composing subgraphs with both Federation 1 and Federation 2. It is recommended that you set `federation_version: 1` or `federation_version: 2` in your YAML configuration file. When Apollo releases new versions of composition for Federation 1 or Federation 2, Rover will find the new package, and download it to your machine. It will then use the new composition package to compose your subgraphs. Our aim is to release only backward-compatible changes across major versions moving forward. If composition breaks from one version to the next, please [submit an issue](https://github.com/apollographql/federation/issues/new?assignees=&labels=&template=bug.md), and follow the instructions for pinning composition to a known version.

Reviewer: ```suggestion
`rover supergraph compose` supports composing subgraphs with both Federation 1 and Federation 2. It is recommended that you set `federation_version: 1` or `federation_version: 2` in your YAML configuration file. When Apollo releases new versions of composition for Federation 1 or Federation 2, Rover finds the new package and downloads it to your machine. It then uses the new composition package to compose your subgraphs. Our aim is to release only backward-compatible changes across major versions moving forward. If composition breaks from one version to the next, please [submit an issue](https://github.com/apollographql/federation/issues/new?assignees=&labels=&template=bug.md), and follow the instructions for pinning composition to a known version.
```

> **⚠️ If you need to pin your composition function to a specific version _(not recommended)_**, you can do so by setting `federation_version: =2.0.1` in your `supergraph.yaml` file. This ensures that Rover _always_ uses the exact version of composition that you specified. In this example, Rover would use `@apollo/composition@v2.0.1`.

>
>
> Versions of the `supergraph` plugin (built from [this source](https://github.com/apollographql/federation-rs)) are installed to `~/.rover/bin` if you installed with the `curl | sh` installer, and to `./node_modules/.bin/` if you installed with npm.

> If you set `federation_version: 1` or `federation_version: 2`, you can run `rover supergraph compose` with the `--skip-update` flag to prevent Rover from downloading newer composition versions. Rover will instead use the latest major version that you have already downloaded to your machine. This can be helpful if you're on a slow network.

Reviewer: ```suggestion
If you set `federation_version: 1` or `federation_version: 2`, you can run `rover supergraph compose` with the `--skip-update` flag to prevent Rover from downloading newer composition versions. Rover instead uses the latest major version that you've downloaded to your machine. This can be helpful if you're on a slow network.
```

> #### Pre v0.5.0 Rover

Reviewer: ```suggestion
#### Earlier Rover versions
```
